### PR TITLE
Feature/tooltip add tests

### PR DESCRIPTION
**Pull request by macfrei**
# _Migration of [tooltip test pr](https://github.com/siva-squad/squadbeyond/pull/746)_

## 概要 / Overview

Researched different options on how to proceed with testing in the new design system. 

1. Split testing and Storybook by exporting stories and running tests with vitest
2. Add tests to stories with the `play` function

Also added unit tests to the utility functions.

新しいデザインシステムでのテストの進め方について、さまざまな選択肢を調査。

1. ストーリーをエクスポートし、vitest でテストを実行することで、スプリットテストと Storybook を行う。
2. `play` 機能でストーリーにテストを追加する。

また、ユーティリティ関数にユニットテストを追加した。

### Jira のチケット / Jira Ticket

https://siva-inc.atlassian.net/browse/DSD-10

## その他 / Other

https://siva-inc.atlassian.net/wiki/spaces/BRUI/pages/125567152/Testing+strategy

### CODE REVIEW

File: src/components/atoms/Tooltip/Tooltip.test.tsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  });
+
+  test("open and closes on hover", async () => {
+    const user = userEvent.setup();
+    render(<Tooltip />);
+
+    const anchor = screen.getByTestId("tooltip-anchor");
+    await user.hover(anchor);
+    const tooltip = screen.getByRole("tooltip");
+    expect(tooltip).toBeInTheDocument();
```
Comment (ogaru-sh) [inline src/components/atoms/Tooltip/Tooltip.test.tsx:25]: `The property 'toBeInTheDocument' is not present in the type 'Assertion<HTMLElement>'.`
This is what my environment displayed
I think I need to add the following to the include of tsconfig.ts?
`  "include": ["src", "vitest.setup.ts"],`

Comment (macfrei) [reply src/components/atoms/Tooltip/Tooltip.test.tsx:25]: Did your try that? I am not getting any messages/errors like that 🤔

Comment (ogaru-sh) [reply src/components/atoms/Tooltip/Tooltip.test.tsx:25]: I wonder why...?
スクリーンショット 2023-11-07 11 54 25

After adding to tsconfig.ts
スクリーンショット 2023-11-07 11 56 01

@yoheeeei @nuintee 
こちらのエラーって表示されますか？

File: src/components/atoms/Tooltip/Tooltip.test.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+// in order to write tests and avoid duplications.
+// Will run with vitest
+import { composeStories } from "@storybook/react";
+import { render, screen, cleanup } from "@testing-library/react";
+import userEvent from "@testing-library/user-event";
+import { describe, expect, test, afterEach } from "vitest";
+
+import * as TooltipStories from "./Tooltip.stories";
+
+const { DefaultOnHover: Tooltip, FocusOnButton } = composeStories(TooltipStories);
```
Comment (ogaru-sh) [inline src/components/atoms/Tooltip/Tooltip.test.tsx:11]: It's great that we can pull this from Storybook!

Comment (macfrei) [reply src/components/atoms/Tooltip/Tooltip.test.tsx:11]: It is! We just need to think about what kind of tests we want to write 🙂

**Comment by ogaru-sh** [comment]
@macfrei 
Personally, given our current state of not having tests implemented yet and the limited number of engineers, I feel that it might be challenging to incorporate both play function tests and Vitest tests for all components in terms of time. 
Also, while I understand that some might argue that writing all tests upfront will make things easier later on, I think it's important to note that not everyone has the same understanding or approach to testing (like agreeing that tests are necessary, but to what extent). 
Therefore, I believe it would be best to start with focusing on one type of test to reduce the burden.

Moreover, as we currently have Visual Tests with Chromatic and reviews from designers, these could also be considered as part of our testing.

As a suggestion, I personally think it might be better to focus on Vitest. This is because Vitest alone can test the component variations in Storybook, and it also offers faster execution speeds.

What are your thoughts on focusing on just one type of test?

個人的には今のまだテストが導入されていない状態かつ少ないエンジニア数という理由から、play関数とvitestの２種類のテストを全てのコンポーネントに入れるのは時間的にも難しいかなという印象ですね。。。
また、最初に全てのテストを書けば後々楽になるという意見もあると思いますが、全員のテストへの感覚（テストはいるけどどこまでやるか）が揃っていないと思うので、一旦は１つのテストに絞って負担が少ないところから始めるのが良いと思いました！
また、今の所ChromaticでVisual Testもあるし、デザイナーレビューもあると思うので、テストとしてカウントできそうと思いました！

提案ですが、個人的にはvitestに絞るのが良いかもと思ってます！
vitestだけでstorybookのコンポーネントバリエーションもテストできて、実行速度も早いので！
Marieさんは１つに絞るところについてどう考えますか？

**Comment by macfrei** [comment]
@ogaru-sh 
I do think we should only use one of the tests, as otherwise it would be quite redundant. I only added both tests because I wanted to showcase that both are possible and do the same thing, just differently 🙂 .

When writing the tests, I enjoyed both ways and I was actually pretty impressed by the `play` function from Storybook. Seeing the step-through-step explanations of the tests in the interactions tab was quite nice and helpful, I think. On the other hand, vitest is quite a bit faster. So yea, I like both approaches and either one is fine! 🙏 

そうしないと冗長になってしまうので、どちらか一方のテストだけを使うべきだと思います。私が両方のテストを追加したのは、どちらも可能であり、同じことをする、ただ違うだけだということを示したかったからです🙂 。

テストを書くとき、私は両方の方法を楽しんだし、実際、Storybookの`play`機能にはかなり感銘を受けた。インタラクション・タブでテストのステップ・スルー・ステップの説明を見ることができるのは、とても親切で役に立ったと思う。一方、vitestはかなり速い。どちらのアプローチも好きだし、どちらでもいい！🙏

**Comment by ogaru-sh** [comment]
But which is actually better...?

@macfrei 
If the purpose of writing tests for a component is to guarantee that it works, as Marie-san wrote,
- `<Tooltip />`  component common test
- `<FocusOnButton />` specific tests

I was wondering if it would be possible to cover this with a relatively minimal number of tests by writing
If so, I think vitest would be a good choice!
(And vitest is fast, haha)

**Comment by ogaru-sh** [comment]
I understood that there are two types of tests!
Thank you!
But which is actually better...?

@macfrei 
If the purpose of writing tests for a component is to guarantee that it works, as Marie-san wrote,
- `<Tooltip />`  component common test
- `<FocusOnButton />` specific tests

I was wondering if it would be possible to cover this with a relatively minimal number of tests by writing
If so, I think vitest would be a good choice!
(And vitest is fast, haha)

**Comment by macfrei** [comment]
@ogaru-sh 
I am not quite sure I understand correctly, I think the translation is a bit off.

The amount of tests doesn't change with the environment you test in.

**Comment by ogaru-sh** [comment]
@macfrei 

Oh, sorry!
I think what I was saying might have been incorrect!

So, whether using the Play function or Vitest, the type of tests we write for the components doesn't change, right?
And the next step is just deciding which one to use, is that correct?

**Comment by macfrei** [comment]
@ogaru-sh 
Yes, that is correct!

**Comment by ogaru-sh** [comment]
@macfrei 
How about focusing exclusively on Vitest? Since tests such as 'util' are also conducted with Vitest, and structurally, tests are separated into tests, it seems likely to become simple and explicit.

File: src/components/atoms/Tooltip/Tooltip.test.tsx (hunk 1 above)
Comment (yoheeeei) [reply src/components/atoms/Tooltip/Tooltip.test.tsx:25]: @ogaru-sh
自分はそのエラー出ませんでしたね...
なんでだろう🤔

**Comment by macfrei** [comment]
@ogaru-sh sounds good, I will delete the tests with the `play` function and remove the dependencies!

**Comment by macfrei** [comment]
@ogaru-sh I removed the play function and the dependencies.

File: src/components/atoms/Tooltip/Tooltip.test.tsx (hunk 1 above)
Comment (ogaru-sh) [reply src/components/atoms/Tooltip/Tooltip.test.tsx:25]: あれ、表示されなくなりました！
大丈夫です！🙏

Oh, it is no longer displayed!
This one is fine!🙏

**Comment by ogaru-sh** [comment]
@macfrei 
LGTM!

**Pull request merged**